src/sqlstorage.cc

```diff
@@ -725,6 +725,26 @@ class SQLTargetWHandle : public StorageTargetWHandle {
       throw exc;
     }
 
+    sqlite3_int64 nfiles;
+    {
+      auto statement =
+          db_.prepareStatement<std::string>("SELECT count(*) FROM target_images WHERE filename = ?;", filename_);
+
+      if (statement.step() != SQLITE_ROW) {
+        LOG_ERROR << "Statement step failure: " << db_.errmsg();
+        throw exc;
+      }
+      nfiles = sqlite3_column_int64(statement.get(), 0);
+    }
+
+    if (nfiles != 0) {
+      int ret = db_.prepareStatement<std::string>("DELETE FROM target_images WHERE filename = ?", filename_).step();
+      if (ret != SQLITE_DONE) {
+        LOG_ERROR << "Statement step failure: " << db_.errmsg();
+        throw exc;
+      }
+    }
+
     auto statement = db_.prepareStatement<std::string, SQLZeroBlob>(
         "INSERT INTO target_images (filename, image_data) VALUES (?, ?);", filename_, SQLZeroBlob{expected_size_});
```

Comment:
@OYTIS  suggests that ON CONFLICT REPLACE might simplify this https://sqlite.org/lang_conflict.html